src/main/scala/wvlet/airframe/AirframeMacros.scala

```diff
@@ -32,37 +32,55 @@ object AirframeMacros extends LogSupport {
     val t = ev.tpe.typeArgs(0)
     val a = t.typeSymbol
 
-    val hasPublicDefaultConstructor = t.members.find(_.isConstructor).map(_.asMethod).exists { m =>
-      m.isPublic && m.paramLists.size == 1 && m.paramLists(0).size == 0
+    // Find the pubilc default constructor without argument list
+    val hasPublicDefaultConstructor = t.members
+                                      .find(_.isConstructor)
+                                      .map(_.asMethod).exists { m =>
+        m.isPublic && m.paramLists.size == 1 && m.paramLists(0).size == 0
+      }
+
+    val shouldInjectSession = if(a.isStatic) {
+      if(!a.isAbstract && hasPublicDefaultConstructor) {
+        true
+      }
+      else {
+        false
+      }
     }
+    else {
+      true
+    }
+
+    // = Abstract type
+    // We cannot build abstract type X, so bind[X].to[ConcreteType]
+    // needs to be found in the design.
+    //
+    // = Non static type
+    // If X is non static type (= local class or trait),
+    // we need to instantiate it first in order to populate its $outer variables
     c.Expr(
-      // = Abstract type
-      // We cannot build abstract type X, so bind[X].to[ConcreteType]
-      // needs to be found in the design.
-      //
-      // = Non static type
-      // If X is non static type (= local class or trait),
-      // we need to instantiate it first in order to populate its $outer variables
-      if(!a.isStatic || (a.isStatic && !a.isAbstract && hasPublicDefaultConstructor)) {
+      if(shouldInjectSession) {
         q"""{
-             ${c.prefix}.getOrElseUpdate[$t]((new $t {
-               protected[this] def __current_session = ${c.prefix}
+          val session = ${c.prefix}
```

Comment:
We need to explicitly bind ${c.prefix} to a variable, because it can be complex chain of method calls.